File: ruby/trinary/trinary.rb
class Trinary
  TRINARY_MATCH = /\A[0-2]+$/

  def initialize(input)
    @input = input
  end

  def to_decimal
    return 0 unless trinary?

    @input
      .split('')
      .reverse
      .each_with_index
      .map { |digit, factor| digit.to_i * 3**factor }
      .reduce(&:+)
  end

  def trinary?
    @input =~ TRINARY_MATCH
  end
end

module BookKeeping
  VERSION = 1
end

File: javascript/trinary/trinary.js
class Trinary {
  constructor(value) {
    this.value = value;
  }

  toDecimal() {
    return this.value.
      replace(/[^012]/g, "").
      split('').
      reverse().
      reduce((total, num, factor) => {
        return total + (parseInt(num, 10) * Math.pow(3, factor))
      }, 0)
  }
}

module.exports = Trinary
